# Лабораторная работа №1
## Выполнила студентка группы БФИ2001 Тимофеева Дарья Валерьевна

### Оглавление
1. [Задание 1](#Задание-№1)
2. [Задание 2](#Задание-№2)
3. [Задание 3](#Задание-№3)
4. [Вывод](#Вывод)

### Задание №1
<i> Вызвать функцию print() и передать туда строку Hello, World! </i>

In [1]:
print('Hello World!')

Hello World!


### Задание №2
Написать генератор случайных матриц(многомерных), который принимает
опциональные параметры <b>m</b>, <b>n</b>, <b>min_limit</b>, <b>max_limit</b>, где <b>m</b> и <b>n</b> указывают размер
матрицы, а <b>min_lim</b> и <b>max_lim</b> - минимальное и максимальное значение для
генерируемого числа.

In [2]:
import random
import time

In [3]:
user_m = int(input('m =  '))
user_n = int(input('n =  '))
user_min_limit = int(input('min_limit =  '))
user_max_limit = int(input('max_limit =  '))
matrix = [ [ random.randint(user_min_limit, user_max_limit) for j in range(user_n)] for i in range(user_m) ]

m =  1000
n =  1000
min_limit =  -250
max_limit =  1250


### Задание №3
Реализовать методы сортировки строк числовой матрицы в соответствии с
заданием. Оценить время работы каждого алгоритма сортировки и сравнить его со
временем стандартной функции сортировки. Испытания проводить на сгенерированных
матрицах.

In [4]:
import copy

In [5]:
# Сортировка выбором.
start_time = time.time()

def selection_sort(mtrx):
    for i in range(len(mtrx)):
        min_index = i
        for j in range(i + 1, len(mtrx)):
            if mtrx[j] < mtrx[min_index]:
                min_index = j
        mtrx[i], mtrx[min_index] = mtrx[min_index], mtrx[i]
    return mtrx

sort_matrix = copy.deepcopy(matrix)
selection_sort(sort_matrix)

print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 338 ms ---


In [6]:
# Сортировка вставкой.
start_time = time.time()
def insertion_sort(mtrx):
    for i in range (1, len(mtrx)):
        insert_el = mtrx[i]
        j = i-1
        while j>=0 and mtrx[j]>insert_el:
            mtrx[j+1] = mtrx[j]
            j-=1
        mtrx[j+1] = insert_el
    return(mtrx)  

sort_matrix = copy.deepcopy(matrix)
insertion_sort(sort_matrix)

print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 330 ms ---


In [7]:
# Сортировка обменом. || Сортировка пузырьком.
start_time = time.time()

def exchange_sort(mtrx):
    for i in range (1, len(mtrx) - 1):
        flag = 0
        for j in range (0, len(mtrx)-1-i):
            if mtrx[j]>=mtrx[j+1]:
                mtrx[j], mtrx[j+1] = mtrx[j+1], mtrx[j]
                flag = 1
        if flag == 0:
            break
    return mtrx

sort_matrix = copy.deepcopy(matrix)
exchange_sort(sort_matrix)

print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 370 ms ---


In [8]:
# Сортировка Шелла.
start_time = time.time()

def shell_sort(mtrx):
    interval = len(mtrx)//2
    while interval>0:
        for i in range(interval, len(mtrx)):
            j = i
            while j>=interval and mtrx[j-interval]>mtrx[i]:
                mtrx[j] = mtrx[j-interval]
                j-=interval
            mtrx[j] = mtrx[i]
        interval//=2
    return(mtrx)

sort_matrix = copy.deepcopy(matrix)
shell_sort(sort_matrix)


print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 318 ms ---


In [9]:
# Быстрая сортировка.
start_time = time.time()

def quick_sort(mtrx):
    if len(mtrx) <= 1:
        return mtrx
    else:
        element = random.choice(mtrx)
        left = []
        middle = []
        right = []
        for i in range(len(mtrx)):
            if mtrx[i] < element:
                left.append(mtrx[i])
            elif mtrx[i] > element:
                right.append(mtrx[i])
            else:
                middle.append(mtrx[i])
        return quick_sort(left) + middle + quick_sort(right)
    
sort_matrix = copy.deepcopy(matrix)
quick_sort(sort_matrix)

print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))
    

--- 307 ms ---


In [10]:
# Турнирная сортировка.
start_time = time.time()

def heapify(mtrx, size, index):
    largest = index
    left_ind = 2 * index + 1
    right_ind = 2 * index + 2
    
    if left_ind < size and mtrx[index] < mtrx[left_ind]:
        largest = left_ind
        
    if right_ind < size and mtrx[largest] < mtrx[right_ind]:
            largest = right_ind
    
    if largest != index:
            mtrx[index], mtrx[largest] = mtrx[largest], mtrx[index]
            heapify(mtrx, size, largest)
            
def tournament_sort(mtrx):
    n = len(mtrx)
    
    for i in range(n - 1, -1, -1):
        heapify(mtrx, n, i)
    
    for k in range(n - 1, 0, -1):
        mtrx[k], mtrx[0] = mtrx[0], mtrx[i] 
        heapify(mtrx, i, 0)

sort_matrix = copy.deepcopy(matrix)
tournament_sort(sort_matrix)

print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 314 ms ---


In [11]:
# Стандартная функция сортировки
start_time = time.time()

sort_matrix = copy.deepcopy(matrix)
sort_matrix.sort()

print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 299 ms ---


### Вывод


         В результате оценки времени работы алгоритмов сортировки массива, состоящего из 1000 элементов, были получены следующие результаты:
         1) Сортировка выбором - 338 ms
         2) Сортировка вставкой - 330 ms
         3) Сортировка обменом - 370 ms
         4) Сортировка Шелла - 318 ms
         5) Быстрая сортировка - 307 ms
         6) Турнирная сортировка - 314 ms
         7) Стандартная функция сортировки - 299 ms
         Стандартная функция сортировки оказалась наиболее быстрой по времени; 
         Наиболее быстрая по времни, за исключением стандартной функции, - быстрая сортировка.
         Наиболее длительная - сортировка обменом.